### DISCUSSION TITLE
How can one implement mostly the same classes

### DISCUSSION THREAD

**Discussion by foreg** [apple/pkl#223] [Q&A] [answered]
Consider an on-premise service offering two subscription plans: Free and Paid. Both plans provide a set of common features as well as some specific. What is the correct way to express `features` property of this service in a template? 

I tried this approach:
```pkl
// template.pkl
open module template

class Feature {
  enabled: Boolean
}

abstract class Features {
  auth: Feature
}

class FreeFeatures extends Features {
  advertisement: Feature
}

class PaidFeatures extends Features {
  exporter: Feature
}

features: FreeFeatures|PaidFeatures
```
```pkl
// free.pkl
extends "./template.pkl"

features {
  auth {
    enabled = true
  }
  advertisement {
    enabled = true
  }
}
```
but the evaluation on `free.pkl` gives me "Tried to read property `features` but its value is undefined."

**Comment by holzensp** [comment, marked as answer]
The default value of a type union is undefined, by default. If you want it to have a default, you have to annotate one of the alternatives with a `*` ([see here](https://pkl-lang.org/main/current/language-reference/index.html#union-types)).

Since `features: FreeFeatures|PaidFeatures` has no default value, there is nothing to amend when you say `features {`. In this case, you should either give it a default value upon its first use (in `free.pkl`):
```
features = new FreeFeatures {
```
of you give it one in the `template.pcl`:
```
features: *FreeFeatures|PaidFeatures
```
Mind you, there is general IDE degradation from using such unions; when I edit `free.pkl`, autocomplete for the body of `features` gives me both `advertisement` and `exporter`.

Also, why is `free.pkl` an `extends` of `template.pcl` and not an `amends`? Your editor experience and overall safety (from typos etc) is much better when you `amends`.

**Reply by foreg** [reply]
Thank you! `features = new FreeFeatures {` is what I was looking for.

> Mind you, there is general IDE degradation from using such unions; when I edit free.pkl, autocomplete for the body of features gives me both advertisement and exporter.

I changed `features`' type to `Features`; it looks much better now

> Also, why is free.pkl an extends of template.pcl and not an amends? Your editor experience and overall safety (from typos etc) is much better when you amends.

I used it to give myself the option to add another arbitrary property unique to free.pkl. What would be a better way to accomplish this?

Thank you once again for your help!
